File: Licensing/Desktop/Distributed/packages/SoftwarePotential.Configuration.Distributor-62340.3.2.1951.6/content/SoftwarePotential/Core/SpAgent.Configuration.Distributor.AppSettings.cs
// NB This file is auto-added via the SoftwarePotential.Configuration.Distributor-<ShortCode> NuGet package.
//
// THE CODE SHOULD BE OVERWRITTEN BY PACKAGE UPDATES - IT IS RECOMMENDED TO DEFINE ANY EXTENSIONS YOU MAY DESIRE ELSEWHERE

using System;
using System.Configuration;
using System.IO;
using System.Text;

namespace Sp.Agent.Distributor.AppSettingsHelpers
{
	/// <summary>
	/// Encapsulates opening (and/or creating and seeding) of a <c>Sp.Agent.Distributor.dll.config</c> 
	/// file that will be maintained outside of an application's <c>app.config</c> file.
	/// </summary>
	class WritableConfigurationFile
	{
		/// <summary>
		/// The name of the config file. As all programmatic access is to be gated 
		/// through this class, this should not need to be externally known.
		/// </summary>
		const string RuntimeConfigFilename = "Sp.Agent.Distributor.dll.config";

		/// <summary>
		/// Opens the file. If the file is not present, places valid stub content there first.
		/// </summary>
		/// <param name="configDirectory">The folder within which the file is to be maintained. Should have appropriate sharing privileges such that all users under whose control the application may need to be reconfigured will have access to write files within it.</param>
		public static SpAgentWritableDistributorAppSettings OpenOrCreateEmpty( string configDirectory )
		{
			return OpenOrInitializeFromTemplateContent( @"<?xml version=""1.0"" encoding=""utf-8"" ?>
<configuration/>", configDirectory );
		}

		static SpAgentWritableDistributorAppSettings OpenOrInitializeFromTemplateContent( string configFileTemplateContent, string configDirectory )
		{
			return OpenOrSeed( runtimeConfigFile => File.WriteAllText( runtimeConfigFile, configFileTemplateContent, Encoding.UTF8 ), configDirectory );
		}

		static SpAgentWritableDistributorAppSettings OpenOrSeed( Action<string> seedFile, string configDirectory )
		{
			var runtimeConfigFile = new FileInfo( Path.Combine( configDirectory, RuntimeConfigFilename ) );
			if ( !runtimeConfigFile.Exists )
				seedFile( runtimeConfigFile.FullName );
			return Open( runtimeConfigFile );
		}

		/// <summary>
		/// Opens the file. Will fail if the file is not present.
		/// </summary>
		/// <param name="runtimeConfigFile">The folder within which the file is stored.</param>
		public static SpAgentWritableDistributorAppSettings Open( FileInfo runtimeConfigFile )
		{
			return SpAgentDistributorConfiguration.OpenStandaloneFile( runtimeConfigFile );
		}
	}

	/// <summary>
	/// Manages opening a Distributor Configuration file from either app.config or a standalone file.
	/// </summary>
	static class SpAgentDistributorConfiguration
	{
		/// <summary>
		/// Accesses [in read/write mode] the BaseUri setting maintained the specified <paramref name="file" />.
		/// </summary>
		public static SpAgentWritableDistributorAppSettings OpenStandaloneFile( FileInfo file )
		{
			return new SpAgentWritableDistributorAppSettings(
				System.Configuration.ConfigurationManager.OpenMappedExeConfiguration(
					new ExeConfigurationFileMap { ExeConfigFilename = file.FullName },
					ConfigurationUserLevel.None ) );
		}

		/// <summary>
		/// Accesses [in read-only mode] the BaseUri setting maintained externally in the application's <c>app.config</c> file.
		/// </summary>
		public static SpAgentDistributorAppSettings FromAppConfig()
		{
			return new SpAgentDistributorAppSettings( System.Configuration.ConfigurationManager.OpenExeConfiguration( ConfigurationUserLevel.None ) );
		}
	}

	/// <summary>
	/// Manages access to an <c>configuration/appSetting/Sp.Agent.Distributor.BaseUri</c> entry in an Application Configuration file.
	/// </summary>
	/// <remarks>
	/// This class handles the reading aspect only. For writing support, see <see cref="SpAgentWritableDistributorAppSettings"/>.
	/// </remarks>
	class SpAgentDistributorAppSettings
	{
		// NB any compiler errors around here mean you're missing an Assembly reference to System.Configuration.dll
		protected readonly System.Configuration.Configuration _configuration;

		/// <summary>
		/// Creates an instance referring to the supplied <see cref="System.Configuration.Configuration"/>.
		/// </summary>
		public SpAgentDistributorAppSettings( System.Configuration.Configuration configuration )
		{
			_configuration = configuration;
		}

		/// <summary>
		/// The setting name the class uses.
		/// </summary>
		/// <value>
		/// <c>Sp.Agent.Distributor.BaseUri</c>
		/// </value>
		protected static string ConfigSettingName
		{
			get { return "Sp.Agent.Distributor.BaseUri"; }
		}

		/// <summary>
		/// Extracts the Configured Uri. Throws if the data is not a valid Uri.
		/// </summary>
		/// <returns>The configured <see cref="Uri"/> or <c>null</c> if no entry is present.</returns>
		/// <exception cref="InvalidOperationException">Thrown if the setting is not a valid <see cref="Uri"/>.</exception>
		public Uri ReadBaseUriOrDefault()
		{
			var setting = _configuration.AppSettings.Settings[ ConfigSettingName ];
			if ( setting == null )
				return null;
			var settingValue = setting.Value;
			Uri uri;
			if ( !Uri.TryCreate( settingValue, UriKind.Absolute, out uri ) )
				throw new InvalidOperationException( "Malformed Uri in Configuration Setting. Has there been an external modification? (" + ConfigSettingName + " = " + settingValue + " in " + Filepath + ")" );
			return uri;
		}

		/// <summary>
		/// The absolute path of the Configuration file we are reading from. 
		/// </summary>
		/// <remarks>
		/// (works regardless of whether we are a .exe.config or a standalone file)
		/// </remarks>
		string Filepath
		{
			get { return _configuration.FilePath; }
		}
	}

	/// <summary>
	/// Manages reading or writing a <c>configuration/appSetting/Sp.Agent.Distributor.BaseUri</c> entry in an Application Configuration file.
	/// </summary>
	/// <remarks>
	/// This class handles the writing aspect (the base class, <see cref="SpAgentDistributorAppSettings"/> handles the reading).
	/// </remarks>
	class SpAgentWritableDistributorAppSettings : SpAgentDistributorAppSettings
	{
		/// <summary>
		/// Creates an instance. The supplied <paramref name="configuration"/> must be writable.
		/// </summary>
		/// <param name="configuration">Configuration. Must be writable.</param>
		public SpAgentWritableDistributorAppSettings( System.Configuration.Configuration configuration )
			: base( configuration )
		{
		}

		/// <summary>
		/// Updates the setting in the associated <see cref="System.Configuration.Configuration"/>.
		/// </summary>
		public void UpdateBaseUri( Uri value )
		{
			if ( value == null )
				throw new ArgumentNullException( "value" );

			_configuration.AppSettings.Settings.Remove( ConfigSettingName );
			_configuration.AppSettings.Settings.Add( ConfigSettingName, value.ToString() );
			_configuration.Save();
		}
	}

	/// <summary>
	/// Caches the (non-<c>null</c> value of writable property. All <c>set</c> operations are expected to be 
	/// passed through the cache too in order to ensure the <c>get</c> will yield a consistent result.
	/// </summary>
	/// <typeparam name="T">The type of value to be cached.</typeparam>
	class WriteThroughCachedValue<T> where T : class
	{
		readonly Action<T> _set;
		readonly Func<T> _get;
		T _cachedValue;

		/// <summary>
		/// Initializes the caching object given a pair of reading/writing delegates.
		/// </summary>
		public WriteThroughCachedValue( Action<T> update, Func<T> retrieve )
		{
			if ( update == null )
				throw new ArgumentNullException( "update" );
			if ( retrieve == null )
				throw new ArgumentNullException( "retrieve" );

			_set = update;
			_get = retrieve;
		}

		/// <summary>
		/// The value being managed.
		/// </summary>
		public T Value
		{
			get
			{
				if ( _cachedValue == null )
					_cachedValue = _get();
				return _cachedValue;
			}
			set
			{
				_set( value );
				_cachedValue = value;
			}
		}
	}
}

File: Licensing/FullFramework/Distributed/packages/SoftwarePotential.Configuration.Distributor-62340.4.0.2022.17/content/SoftwarePotential.Vb/Core/SpAgent.Configuration.Distributor.AppSettings.vb
' NB This file is auto-added via the SoftwarePotential.Configuration.Distributor-<ShortCode> NuGet package.
'
' THE CODE SHOULD BE OVERWRITTEN BY PACKAGE UPDATES - IT IS RECOMMENDED TO DEFINE ANY EXTENSIONS YOU MAY DESIRE ELSEWHERE

Imports System
Imports System.Configuration

''' <summary>
''' Encapsulates opening (and/or creating and seeding) of a <c>Sp.Agent.Distributor.dll.config</c> 
''' file that will be maintained outside of an application's <c>app.config</c> file.
''' </summary>
Class WritableConfigurationFile
	Private Const RuntimeConfigFilename As String = "Sp.Agent.Distributor.dll.config"

	''' <summary>
	''' The name of the config file. As all programmatic access is to be gated 
	''' through this class, this should not need to be externally known.
	''' </summary>
	Public Shared Function OpenOrCreateEmpty(configDirectory As String) As SpAgentWritableDistributorAppSettings
		Return WritableConfigurationFile.OpenOrInitializeFromTemplateContent("<?xml version=""1.0"" encoding=""utf-8"" ?>" & Environment.NewLine & "<configuration/>", configDirectory)
	End Function

	Private Shared Function OpenOrInitializeFromTemplateContent(configFileTemplateContent As String, configDirectory As String) As SpAgentWritableDistributorAppSettings
		Return WritableConfigurationFile.OpenOrSeed(Sub(runtimeConfigFile As String)
			System.IO.File.WriteAllText(runtimeConfigFile, configFileTemplateContent, System.Text.Encoding.UTF8)
																End Sub, configDirectory)
	End Function

	Private Shared Function OpenOrSeed(seedFile As System.Action(Of String), configDirectory As String) As SpAgentWritableDistributorAppSettings
		Dim runtimeConfigFile As System.IO.FileInfo = New System.IO.FileInfo(System.IO.Path.Combine(configDirectory, "Sp.Agent.Distributor.dll.config"))
		If Not runtimeConfigFile.Exists Then
			seedFile(runtimeConfigFile.FullName)
		End If
		Return WritableConfigurationFile.Open(runtimeConfigFile)
	End Function

	''' <summary>
	''' Opens the file. Will fail if the file is not present.
	''' </summary>
	''' <param name="runtimeConfigFile">The folder within which the file is stored.</param>
	Public Shared Function Open(runtimeConfigFile As System.IO.FileInfo) As SpAgentWritableDistributorAppSettings
		Return SpAgentDistributorConfiguration.OpenStandaloneFile(runtimeConfigFile)
	End Function
End Class

Class SpAgentDistributorAppSettings
	Protected _configuration As System.Configuration.Configuration

	Protected Shared ReadOnly Property ConfigSettingName() As String
		Get
			Return "Sp.Agent.Distributor.BaseUri"
		End Get
	End Property

	Private ReadOnly Property Filepath() As String
		Get
			Return Me._configuration.FilePath
		End Get
	End Property

	Public Sub New(configuration As System.Configuration.Configuration)
		Me._configuration = configuration
	End Sub

	Public Function ReadBaseUriOrDefault() As Uri
		Dim setting As KeyValueConfigurationElement = Me._configuration.AppSettings.Settings(SpAgentDistributorAppSettings.ConfigSettingName)
		Dim result As Uri
		If setting Is Nothing Then
			result = Nothing
		Else
			Dim settingValue As String = setting.Value
			Dim uri As Uri = Nothing
			If Not uri.TryCreate(settingValue, UriKind.Absolute, uri) Then
				Throw New System.InvalidOperationException(String.Concat( "Malformed Uri in Configuration Setting. Has there been an external modification? (", SpAgentDistributorAppSettings.ConfigSettingName, " = ", settingValue, " in ", Me.Filepath, ")" ))
			End If
			result = uri
		End If
		Return result
	End Function
End Class

''' <summary>
''' Manages opening a Distributor Configuration file from either app.config or a standalone file.
''' </summary>
Module SpAgentDistributorConfiguration
	''' <summary>
	''' Accesses [in read/write mode] the BaseUri setting maintained the specified <paramref name="file" />.
	''' </summary>
	Public Function OpenStandaloneFile(file As System.IO.FileInfo) As SpAgentWritableDistributorAppSettings
		Return New SpAgentWritableDistributorAppSettings(ConfigurationManager.OpenMappedExeConfiguration(New ExeConfigurationFileMap() With {.ExeConfigFilename = file.FullName}, ConfigurationUserLevel.None))
	End Function

	''' <summary>
	''' Accesses [in read-only mode] the BaseUri setting maintained externally in the application's <c>app.config</c> file.
	''' </summary>
	Public Function FromAppConfig() As SpAgentDistributorAppSettings
		Return New SpAgentDistributorAppSettings(ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None))
	End Function
End Module

''' <summary>
''' Manages access to an <c>configuration/appSetting/Sp.Agent.Distributor.BaseUri</c> entry in an Application Configuration file.
''' </summary>
''' <remarks>
''' This class handles the reading aspect only. For writing support, see <see cref="SpAgentWritableDistributorAppSettings"/>.
''' </remarks>
Class SpAgentWritableDistributorAppSettings
	Inherits SpAgentDistributorAppSettings

	Public Sub New(configuration As System.Configuration.Configuration)
		MyBase.New(configuration)
	End Sub

	Public Sub UpdateBaseUri(value As Uri)
		Me._configuration.AppSettings.Settings.Remove(SpAgentDistributorAppSettings.ConfigSettingName)

		If value IsNot Nothing Then
			Me._configuration.AppSettings.Settings.Add(SpAgentDistributorAppSettings.ConfigSettingName, value.ToString())
		End If
		
		Me._configuration.Save()
	End Sub
End Class

''' <summary>
''' Caches the (non-<c>null</c> value of writable property. All <c>set</c> operations are expected to be 
''' passed through the cache too in order to ensure the <c>get</c> will yield a consistent result.
''' </summary>
''' <typeparam name="T">The type of value to be cached.</typeparam>
Class WriteThroughCachedValue(Of T)
	Private _set As System.Action(Of T)

	Private _get As Func(Of T)

	Private _cachedValue As T

	Public Property Value() As T
		Get
			If Me._cachedValue Is Nothing Then
				Me._cachedValue = Me._get()
			End If
			Return Me._cachedValue
		End Get
		Set(value As T)
			Me._set(value)
			Me._cachedValue = value
		End Set
	End Property

	Public Sub New(update As System.Action(Of T), retrieve As Func(Of T))
		If update Is Nothing Then
			Throw New System.ArgumentNullException("update")
		End If
		If retrieve Is Nothing Then
			Throw New System.ArgumentNullException("retrieve")
		End If
		Me._set = update
		Me._get = retrieve
	End Sub
End Class
